### PR TITLE
Fix running command line with more than a single word under MSW

### LINKED ISSUE #18: `run` on Windows appears to escape quotes
_Originally posted by @vadz in https://github.com/tridactyl/tridactyl/issues/876#issuecomment-766111592_

There seems to be something wrong with command line parsing: my `editorcmd` is `gvim.exe -f -c "se ft=markdown wrap lbr tw=0" %f` and it worked fine with 0.1.10 before, but now it starts editing the file called `ft=markdown` and a few more because the quotes get escaped and I couldn't find any way to prevent this from happening (other than removing `-c ...` entirely). Would you have any idea about this by chance? TIA!

---

https://github.com/tridactyl/native_messenger/blob/100198fb4c1e5167ad59d310c93d4d3ef8e77ee8/src/native_main.nim#L108

### PR DESCRIPTION
**Pull request by vadz**
Split command line into arguments when passing it to cmd.exe, respecting
quotes (but not anything else for now).

Closes #18.

---

Sorry for my baby Nim, I'm pretty sure this could be done in some better way, but at least this works for me (in fact, I'm writing this text in Vim, opened from Firefox) and is already an improvement.

### CODE REVIEW

File: src/native_main.nim [hunk 1]
```diff
@@ -37,6 +37,36 @@ proc sanitiseFilename(fn: string): string =
...
+        case c:
+            of ' ':
+                if inQuotes:
+                    arg.add(c)
+                else:
+                    if arg != "":
+                        args.add(arg)
+                        arg = ""
+
+            of '"':
```
Comment (bovine3dom) [inline src/native_main.nim:59]:
```suggestion
            of '"', '\'':
```

I think cmd treats the two interchangeably?

Comment (vadz) [reply src/native_main.nim:59]: No, `'` is not special to it. Just try running `gvim.exe -c "set ft=markdown"` and `gvim.exe -c 'set ft=markdown'` from its prompt.

**Comment by bovine3dom** [comment]
Thanks for this.

I'm _really_ reluctant to merge this because I can see loads of edge cases - escaped quotes, nested quotes - that will cause weird behaviour.

I don't think it should be too difficult to do it "properly", e.g. with `os.quoteShell` or the dumb "prepend command with cmd /c and then do the same as we do in Linux" I suggested.

If we can't do that then, yeah, I agree that this is better than the current situation for users.

File: src/native_main.nim (hunk 1 above)
Comment (Rummskartoffel) [reply src/native_main.nim:59]: I don't think it does – as far as I understand, single quotes are treated like normal characters: see [this SO answer](https://stackoverflow.com/a/51090074/9859881).

**Comment by Rummskartoffel** [comment]
I presume this would break with things like `somecommand "foo """bar""""`. While I think you're just asking to suffer once you start trying to fight cmd's quoting rules, this still wouldn't be backwards compatible with the old messenger, then.
On another note, I did some googling again and I think `poEvalCommand` doesn't mean "shell" to refer to cmd, but the thing that pre-processes and constructs `argv`, which has [another different set of rules](https://docs.microsoft.com/en-us/cpp/cpp/main-function-command-line-args). I think that's what `os.quoteShell` is meant for and what `startProcess` directly passes its `command` to when `poEvalCommand` is used.

All of this is of course under the assumption that I even understand anything here correctly, which I'm not sure about.

File: src/native_main.nim (hunk 1 above)
Comment (bovine3dom) [reply src/native_main.nim:59]: > No, ' is not special to it.

Thanks : )

> Just try running gvim.exe -c "set ft=markdown" and gvim.exe -c 'set ft=markdown' from its prompt

If I had access to a Windows machine I probably would have fixed this myself :P

Comment (vadz) [reply src/native_main.nim:59]: FWIW you can use Wine for simple tests like this (but you probably don't want to run FF under Wine...).

**Comment by vadz** [comment]
> I'm _really_ reluctant to merge this because I can see loads of edge cases - escaped quotes, nested quotes - that will cause weird behaviour.

I could add support for using backslashes to escape quotes, if it could help.
 
> I don't think it should be too difficult to do it "properly", e.g. with `os.quoteShell` or the dumb "prepend command with cmd /c and then do the same as we do in Linux" I suggested.

`quoteShell()` seems to work in the other direction, i.e. it takes a single argument and quotes it if necessary. We need to split a string into arguments here and AFAICS Nim doesn't provide any support for this. As for `cmd /c`, isn't this what we already do? If not, could you please explain what do you mean?

Anyhow, I'd just like to emphasize that the current version, which passes the entire string of the command as a single elements of `args` is broken. This string must be split into individual arguments in some way.

**Comment by bovine3dom** [comment]
On Linux we use poEvalCommand which doesn't do any escaping. I'm suggesting that we do the same thing on Windows, which is what we did initially, but add cmd /c to the start of the command.

The reason we stopped using it on Windows is because poEvalCommand dumps you in a shell without any built-ins (e.g. which).

With poEvalCommand you give startProcess a single string and don't use the arguments array.

I think you're probably right and quoteshell does the opposite of what we want...
----
PGP - 2BB8 C36B B504 BFF3 - https://keybase.io/bovine3dom/key.asc

**Comment by Rummskartoffel** [comment]
If all else fails, we could use [`_popen` or `_wpopen`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/popen-wpopen) – probably the latter, for Unicode support –, but that'd mean more and even uglier platform-specific code.

**Comment by vadz** [comment]
OK, I can confirm that using `poEvalCommand` works as well, please see #21, thanks for the idea.
